##**DEEP DATASET "MAVEN FUZZY FACTORY"**

## #TABLE WEBSITE_SESSIONS

In [ ]:
import pandas as pd

In [ ]:
potter_url = 'https://docs.google.com/spreadsheets/d/1xdSUObWQwSIx2EFhmJH2o9Otbu7ei4TghvvVcsmuLl0/edit?gid=125658187#gid=125658187'
spreadsheet_id = potter_url.split('/')[-2]
gid = potter_url.split('gid=')[-1].split('#')[0]
potter_csv_export_url = f'https://docs.google.com/spreadsheets/d/{spreadsheet_id}/export?format=csv&gid={gid}'
df_website_sessions = pd.read_csv(potter_csv_export_url)
df_website_sessions

website_session_id    created_at  user_id  is_repeat_session  \
0                        1  3/19/12 8:04        1                  0   
1                        2  3/19/12 8:16        2                  0   
2                        3  3/19/12 8:26        3                  0   
3                        4  3/19/12 8:37        4                  0   
4                        5  3/19/12 9:00        5                  0   
...                    ...           ...      ...                ...   
472866              472867  3/19/15 7:49   394314                  0   
472867              472868  3/19/15 7:54   394315                  0   
472868              472869  3/19/15 7:55   394316                  0   
472869              472870  3/19/15 7:56   394317                  0   
472870              472871  3/19/15 7:59   394318                  0   

       utm_source utm_campaign utm_content device_type  \
0         gsearch     nonbrand      g_ad_1      mobile   
1         gsearch     nonbrand      g_ad_1     desktop   
2         gsearch     nonbrand      g_ad_1     desktop   
3         gsearch     nonbrand      g_ad_1     desktop   
4         gsearch     nonbrand      g_ad_1      mobile   
...           ...          ...         ...         ...   
472866    gsearch        brand      g_ad_2     desktop   
472867    bsearch     nonbrand      b_ad_1      mobile   
472868    gsearch     nonbrand      g_ad_1      mobile   
472869    gsearch     nonbrand      g_ad_1     desktop   
472870        NaN          NaN         NaN      mobile   

                   http_referer  
0       https://www.gsearch.com  
1       https://www.gsearch.com  
2       https://www.gsearch.com  
3       https://www.gsearch.com  
4       https://www.gsearch.com  
...                         ...  
472866  https://www.gsearch.com  
472867  https://www.bsearch.com  
472868  https://www.gsearch.com  
472869  https://www.gsearch.com  
472870                      NaN  

[472871 rows x 9 columns]

In [ ]:
backup1=df_website_sessions.copy()

In [ ]:
df_website_sessions.head()

website_session_id    created_at  user_id  is_repeat_session utm_source  \
0                   1  3/19/12 8:04        1                  0    gsearch   
1                   2  3/19/12 8:16        2                  0    gsearch   
2                   3  3/19/12 8:26        3                  0    gsearch   
3                   4  3/19/12 8:37        4                  0    gsearch   
4                   5  3/19/12 9:00        5                  0    gsearch   

  utm_campaign utm_content device_type             http_referer  
0     nonbrand      g_ad_1      mobile  https://www.gsearch.com  
1     nonbrand      g_ad_1     desktop  https://www.gsearch.com  
2     nonbrand      g_ad_1     desktop  https://www.gsearch.com  
3     nonbrand      g_ad_1     desktop  https://www.gsearch.com  
4     nonbrand      g_ad_1      mobile  https://www.gsearch.com

In [ ]:
df_website_sessions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 472871 entries, 0 to 472870
Data columns (total 9 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   website_session_id  472871 non-null  int64 
 1   created_at          472871 non-null  object
 2   user_id             472871 non-null  int64 
 3   is_repeat_session   472871 non-null  int64 
 4   utm_source          389543 non-null  object
 5   utm_campaign        389543 non-null  object
 6   utm_content         389543 non-null  object
 7   device_type         472871 non-null  object
 8   http_referer        432954 non-null  object
dtypes: int64(3), object(6)
memory usage: 32.5+ MB


ada data yang bernilai Null dibeberapa column, diantaranya : 1. utm_source (389543 non Null) --> 83328 Null,
2. utm_campaign (389543 non Null) --> 83328 Null,
3. utm_content (389543 non Null) --> 83328 Null,
4. http_referer (432954 non Null) --> 39917 Null

##CEK TIPE DATA

In [ ]:
df_website_sessions.dtypes

website_session_id     int64
created_at            object
user_id                int64
is_repeat_session      int64
utm_source            object
utm_campaign          object
utm_content           object
device_type           object
http_referer          object
dtype: object

tipe data di column created_at perlu diubah dr object ke datetime

##CEK TYPO

In [ ]:
for column in df_website_sessions:
    print(df_website_sessions[column].value_counts())

website_session_id
472871    1
1         1
2         1
3         1
4         1
         ..
12        1
11        1
10        1
9         1
8         1
Name: count, Length: 472871, dtype: int64
created_at
11/28/14 9:39     14
11/28/14 12:08    13
11/29/13 13:00    13
11/29/13 12:45    13
11/28/14 15:31    12
                  ..
3/19/15 7:28       1
3/19/15 7:29       1
3/19/12 8:37       1
3/19/12 8:26       1
3/19/12 8:16       1
Name: count, Length: 367237, dtype: int64
user_id
96978     4
96977     4
322408    4
247449    4
254911    4
         ..
138157    1
138155    1
138153    1
138152    1
138166    1
Name: count, Length: 394318, dtype: int64
is_repeat_session
0    394318
1     78553
Name: count, dtype: int64
utm_source
gsearch       316035
bsearch        62823
socialbook     10685
Name: count, dtype: int64
utm_campaign
nonbrand            337615
brand                41243
desktop_targeted      5590
pilot                 5095
Name: count, dtype: int64
utm_content
g_ad_1        

In [ ]:
df_website_sessions['utm_source'].value_counts()

utm_source
gsearch       316035
bsearch        62823
socialbook     10685
Name: count, dtype: int64

In [ ]:
df_website_sessions['utm_campaign'].value_counts()

utm_campaign
nonbrand            337615
brand                41243
desktop_targeted      5590
pilot                 5095
Name: count, dtype: int64

In [ ]:
df_website_sessions['utm_content'].value_counts()

utm_content
g_ad_1         282706
b_ad_1          54909
g_ad_2          33329
b_ad_2           7914
social_ad_2      5590
social_ad_1      5095
Name: count, dtype: int64

In [ ]:
df_website_sessions['device_type'].value_counts()

device_type
desktop    327027
mobile     145844
Name: count, dtype: int64

setelah pengecekan dengan .value_counts() tdk ditemukan adanya TYPO di semua column pada table website_sessions

##CEK DUPLICATE

In [ ]:
df_website_sessions['website_session_id'].duplicated()

0         False
1         False
2         False
3         False
4         False
          ...  
472866    False
472867    False
472868    False
472869    False
472870    False
Name: website_session_id, Length: 472871, dtype: bool

setelah pengecekan data duplicate, tdk ditemukan adanya data yg duplicate di column website_session_id, sedangkan untuk column user_id, duplicate data diperbolehkan krn berdasarkan data dictionary, user_id di table website_session bukanlah primery key (PK) yg harus bernilai unik.

##CEK MISSING VALUE

In [ ]:
df_website_sessions.isnull().sum()

website_session_id        0
created_at                0
user_id                   0
is_repeat_session         0
utm_source            83328
utm_campaign          83328
utm_content           83328
device_type               0
http_referer          39917
dtype: int64

adanya missing value di column utm_source, utm_campaign, utm_content dan http_referer

## DATA CLEANING AND PREPARATION

###merubah tipe data di column created_at dari object menjadi datetime

In [ ]:
df_website_sessions['created_at'] = pd.to_datetime(df_website_sessions['created_at'], format='%m/%d/%y %H:%M')

In [ ]:
df_website_sessions.dtypes

website_session_id             int64
created_at            datetime64[ns]
user_id                        int64
is_repeat_session              int64
utm_source                    object
utm_campaign                  object
utm_content                   object
device_type                   object
http_referer                  object
dtype: object

###melakukan pengisian pada column yg missing value yaitu pada column utm_source, utm_campaign, utm_content dan http_referer

In [ ]:
df_website_sessions['utm_source'].fillna('direct', inplace=True)
df_website_sessions['utm_campaign'].fillna('non_campaign', inplace=True)
df_website_sessions['utm_content'].fillna('generic', inplace=True)
df_website_sessions['http_referer'].fillna('direct', inplace=True)

/tmp/ipython-input-117010731.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_website_sessions['utm_source'].fillna('direct', inplace=True)
/tmp/ipython-input-117010731.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=

In [ ]:
df_website_sessions.isnull().sum()

website_session_id    0
created_at            0
user_id               0
is_repeat_session     0
utm_source            0
utm_campaign          0
utm_content           0
device_type           0
http_referer          0
dtype: int64

## #TABLE PRODUCTS

In [ ]:
import pandas as pd

In [ ]:
potter_url = 'https://docs.google.com/spreadsheets/d/1-M8lhvA0HNPoBJNqL5CuW1TfLkxIKHgXm2EeSN9RFTM/edit?gid=1571018562#gid=1571018562'
spreadsheet_id = potter_url.split('/')[-2]
gid = potter_url.split('gid=')[-1].split('#')[0]
potter_csv_export_url = f'https://docs.google.com/spreadsheets/d/{spreadsheet_id}/export?format=csv&gid={gid}'
df_products = pd.read_csv(potter_csv_export_url)
df_products

product_id     created_at                product_name
0           1   3/19/12 8:00      The Original Mr. Fuzzy
1           2   1/6/13 13:00       The Forever Love Bear
2           3  12/12/13 9:00    The Birthday Sugar Panda
3           4   2/5/14 10:00  The Hudson River Mini bear

In [ ]:
backup1=df_products.copy()

##CEK TIPE DATA

In [ ]:
df_products.dtypes

product_id       int64
created_at      object
product_name    object
dtype: object

tipe data di column created_at perlu diubah dr object ke datetime

In [ ]:
df_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   product_id    4 non-null      int64 
 1   created_at    4 non-null      object
 2   product_name  4 non-null      object
dtypes: int64(1), object(2)
memory usage: 228.0+ bytes


tidak ditemukan data di column yg bernilai Null

##CEK TYPO

In [ ]:
for column in df_products:
    print(df_products[column].value_counts())

product_id
1    1
2    1
3    1
4    1
Name: count, dtype: int64
created_at
3/19/12 8:00     1
1/6/13 13:00     1
12/12/13 9:00    1
2/5/14 10:00     1
Name: count, dtype: int64
product_name
The Original Mr. Fuzzy        1
The Forever Love Bear         1
The Birthday Sugar Panda      1
The Hudson River Mini bear    1
Name: count, dtype: int64


setelah pengecekan dengan .value_counts() tdk ditemukan adanya TYPO di semua column pada table products

##CEK DUPLICATE

In [ ]:
df_products['product_id'].duplicated()

0    False
1    False
2    False
3    False
Name: product_id, dtype: bool

In [ ]:
df_products['created_at'].duplicated()

0    False
1    False
2    False
3    False
Name: created_at, dtype: bool

In [ ]:
df_products['product_name'].duplicated()

0    False
1    False
2    False
3    False
Name: product_name, dtype: bool

setelah dilakukan pengecekan dengan .duplicated, tdk ditemukan adanya data yg duplicate di semua column pada tabel products

##CEK MISSING VALUE

In [ ]:
df_products.isnull().sum()

product_id      0
created_at      0
product_name    0
dtype: int64

setelah dilakukan pengecekan missing value dengan .isnull().sum(), tidak ditemukan adanya data yg missing value di semua column pada tabel products

##DATA CLEANING AND PREPARATION

###merubah tipe data di column created_at dari object menjadi datetime

In [ ]:
df_products['created_at'] = pd.to_datetime(df_products['created_at'], format='%m/%d/%y %H:%M')

In [ ]:
df_products.dtypes

product_id               int64
created_at      datetime64[ns]
product_name            object
dtype: object

## #TABLE WEBSITE_PAGEVIEWS

In [ ]:
import pandas as pd

In [ ]:
potter_url = 'https://docs.google.com/spreadsheets/d/1DGSUZmY0fOSqLeghEnRQ8hd-M-TipYrZdQEAaF9xXPo/edit?gid=1050705122#gid=1050705122'
spreadsheet_id = potter_url.split('/')[-2]
gid = potter_url.split('gid=')[-1].split('#')[0]
potter_csv_export_url = f'https://docs.google.com/spreadsheets/d/{spreadsheet_id}/export?format=csv&gid={gid}'
df_website_pageviews = pd.read_csv(potter_csv_export_url)
df_website_pageviews

website_pageview_id     created_at  website_session_id  \
0                          1   3/19/12 8:04                   1   
1                          2   3/19/12 8:16                   2   
2                          3   3/19/12 8:26                   3   
3                          4   3/19/12 8:37                   4   
4                          5   3/19/12 9:00                   5   
...                      ...            ...                 ...   
1048570              1048571  1/16/15 18:59              421952   
1048571              1048572  1/16/15 19:01              421954   
1048572              1048573  1/16/15 19:03              421952   
1048573              1048574  1/16/15 19:03              421956   
1048574              1048575  1/16/15 19:03              421952   

                      pageview_url  
0                            /home  
1                            /home  
2                            /home  
3                            /home  
4                            /home  
...                            ...  
1048570  /the-birthday-sugar-panda  
1048571                  /products  
1048572                      /cart  
1048573                      /home  
1048574                  /shipping  

[1048575 rows x 4 columns]

In [ ]:
backup1=df_website_pageviews.copy()

In [ ]:
df_website_pageviews.head()

website_pageview_id    created_at  website_session_id pageview_url
0                    1  3/19/12 8:04                   1        /home
1                    2  3/19/12 8:16                   2        /home
2                    3  3/19/12 8:26                   3        /home
3                    4  3/19/12 8:37                   4        /home
4                    5  3/19/12 9:00                   5        /home

In [ ]:
df_website_pageviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 4 columns):
 #   Column               Non-Null Count    Dtype 
---  ------               --------------    ----- 
 0   website_pageview_id  1048575 non-null  int64 
 1   created_at           1048575 non-null  object
 2   website_session_id   1048575 non-null  int64 
 3   pageview_url         1048575 non-null  object
dtypes: int64(2), object(2)
memory usage: 32.0+ MB


tidak ditemukan data di column yg bernilai Null

##CEK TIPE DATA

In [ ]:
df_website_pageviews.dtypes

website_pageview_id     int64
created_at             object
website_session_id      int64
pageview_url           object
dtype: object

tipe data di column created_at perlu diubah dr object ke datetime

##CEK TYPO

In [ ]:
for column in df_website_pageviews:
    print(df_website_pageviews[column].value_counts())

website_pageview_id
1048575    1
1          1
2          1
3          1
4          1
          ..
20         1
19         1
18         1
17         1
16         1
Name: count, Length: 1048575, dtype: int64
created_at
11/28/14 14:21    27
11/28/14 10:30    26
11/28/14 15:47    25
11/28/14 10:34    25
11/28/14 15:52    25
                  ..
1/16/15 18:44      1
1/16/15 18:43      1
1/16/15 18:40      1
1/16/15 18:37      1
1/16/15 18:36      1
Name: count, Length: 615935, dtype: int64
website_session_id
230520    7
421928    7
20        7
230547    7
380508    7
         ..
230518    1
8         1
9         1
10        1
11        1
Name: count, Length: 421956, dtype: int64
pageview_url
/products                      230534
/the-original-mr-fuzzy         145884
/lander-2                      131170
/home                          119993
/cart                           82603
/lander-3                       69969
/shipping                       56124
/lander-1                       47574


In [ ]:
df_website_pageviews['website_pageview_id'].value_counts()

website_pageview_id
1048575    1
1          1
2          1
3          1
4          1
          ..
20         1
19         1
18         1
17         1
16         1
Name: count, Length: 1048575, dtype: int64

In [ ]:
df_website_pageviews['created_at'].value_counts()

created_at
11/28/14 14:21    27
11/28/14 10:30    26
11/28/14 15:47    25
11/28/14 10:34    25
11/28/14 15:52    25
                  ..
1/16/15 18:44      1
1/16/15 18:43      1
1/16/15 18:40      1
1/16/15 18:37      1
1/16/15 18:36      1
Name: count, Length: 615935, dtype: int64

In [ ]:
df_website_pageviews['website_session_id'].value_counts()

website_session_id
230520    7
421928    7
20        7
230547    7
380508    7
         ..
230518    1
8         1
9         1
10        1
11        1
Name: count, Length: 421956, dtype: int64

In [ ]:
df_website_pageviews['pageview_url'].value_counts()

pageview_url
/products                      230534
/the-original-mr-fuzzy         145884
/lander-2                      131170
/home                          119993
/cart                           82603
/lander-3                       69969
/shipping                       56124
/lander-1                       47574
/lander-5                       43865
/billing-2                      41696
/thank-you-for-your-order       27972
/the-forever-love-bear          21410
/the-birthday-sugar-panda       15690
/lander-4                        9385
/billing                         3617
/the-hudson-river-mini-bear      1089
Name: count, dtype: int64

setelah pengecekan dengan .value_counts() tdk ditemukan adanya TYPO di semua column pada table website_pageviews

##CEK DUPLICATE

In [ ]:
df_website_pageviews['website_pageview_id'].duplicated()

0          False
1          False
2          False
3          False
4          False
           ...  
1048570    False
1048571    False
1048572    False
1048573    False
1048574    False
Name: website_pageview_id, Length: 1048575, dtype: bool

In [ ]:
df_website_pageviews['created_at'].duplicated()

0          False
1          False
2          False
3          False
4          False
           ...  
1048570    False
1048571    False
1048572    False
1048573     True
1048574     True
Name: created_at, Length: 1048575, dtype: bool

In [ ]:
df_website_pageviews['website_session_id'].duplicated()

0          False
1          False
2          False
3          False
4          False
           ...  
1048570     True
1048571     True
1048572     True
1048573    False
1048574     True
Name: website_session_id, Length: 1048575, dtype: bool

In [ ]:
df_website_pageviews['pageview_url'].duplicated()

0          False
1           True
2           True
3           True
4           True
           ...  
1048570     True
1048571     True
1048572     True
1048573     True
1048574     True
Name: pageview_url, Length: 1048575, dtype: bool

setelah dilakukan pengecekan dengan .duplicated, tdk ditemukan adanya data yg duplicate di semua column pada tabel website_pageviews

##CEK MISSING VALUE

In [ ]:
df_website_pageviews.isnull().sum()

website_pageview_id    0
created_at             0
website_session_id     0
pageview_url           0
dtype: int64

setelah dilakukan pengecekan missing value dengan .isnull().sum(), tidak ditemukan adanya data yg missing value di semua column pada tabel website_pageviews

##DATA CLEANING AND PREPARATION

###merubah tipe data di column created_at dari object menjadi datetime

In [ ]:
df_website_pageviews['created_at'] = pd.to_datetime(df_website_pageviews['created_at'], format='%m/%d/%y %H:%M')

In [ ]:
df_website_pageviews.dtypes

website_pageview_id             int64
created_at             datetime64[ns]
website_session_id              int64
pageview_url                   object
dtype: object

In [ ]:
df_website_pageviews.head()

website_pageview_id          created_at  website_session_id pageview_url
0                    1 2012-03-19 08:04:00                   1        /home
1                    2 2012-03-19 08:16:00                   2        /home
2                    3 2012-03-19 08:26:00                   3        /home
3                    4 2012-03-19 08:37:00                   4        /home
4                    5 2012-03-19 09:00:00                   5        /home

###conversi semua huruf di URL di column pegeview_url menjadi lowercase untuk memudahkan pencocokan string.

In [ ]:
product_names = df_products['product_name'].str.lower().tolist()
df_website_pageviews["pageview_url_lower"] = df_website_pageviews["pageview_url"].str.lower()
df_website_pageviews["is_product_view"] = df_website_pageviews["pageview_url_lower"].apply(
    lambda url: any(product in url for product in product_names)
)

In [ ]:
df_website_pageviews[["pageview_url", "is_product_view"]].head()

pageview_url  is_product_view
0        /home            False
1        /home            False
2        /home            False
3        /home            False
4        /home            False

## #TABLE ORDERS

In [ ]:
import pandas as pd

In [ ]:
potter_url = 'https://docs.google.com/spreadsheets/d/10YemLXrRqoTeo3CzjBbY2DddfHGrUD0Y8R4c9LU6q_8/edit?gid=1097446438#gid=1097446438'
spreadsheet_id = potter_url.split('/')[-2]
gid = potter_url.split('gid=')[-1].split('#')[0]
potter_csv_export_url = f'https://docs.google.com/spreadsheets/d/{spreadsheet_id}/export?format=csv&gid={gid}'
df_orders = pd.read_csv(potter_csv_export_url)
df_orders

order_id     created_at  website_session_id  user_id  \
0             1  3/19/12 10:42                  20       20   
1             2  3/19/12 19:27                 104      104   
2             3   3/20/12 6:44                 147      147   
3             4   3/20/12 9:41                 160      160   
4             5  3/20/12 11:28                 177      177   
...         ...            ...                 ...      ...   
32308     32309   3/19/15 3:58              472795   394255   
32309     32310   3/19/15 4:10              472798   394257   
32310     32311   3/19/15 5:27              472809   394268   
32311     32312   3/19/15 5:35              472814   394273   
32312     32313   3/19/15 5:38              472818   386000   

       primary_product_id  items_purchased price_usd cogs_usd  
0                       1                1     49,99    19,49  
1                       1                1     49,99    19,49  
2                       1                1     49,99    19,49  
3                       1                1     49,99    19,49  
4                       1                1     49,99    19,49  
...                   ...              ...       ...      ...  
32308                   1                1     49,99    19,49  
32309                   4                1     29,99     9,49  
32310                   2                2     89,98    31,98  
32311                   4                1     29,99     9,49  
32312                   1                1     49,99    19,49  

[32313 rows x 8 columns]

In [ ]:
backup1=df_orders.copy()

In [ ]:
df_orders.head()

order_id     created_at  website_session_id  user_id  primary_product_id  \
0         1  3/19/12 10:42                  20       20                   1   
1         2  3/19/12 19:27                 104      104                   1   
2         3   3/20/12 6:44                 147      147                   1   
3         4   3/20/12 9:41                 160      160                   1   
4         5  3/20/12 11:28                 177      177                   1   

   items_purchased price_usd cogs_usd  
0                1     49,99    19,49  
1                1     49,99    19,49  
2                1     49,99    19,49  
3                1     49,99    19,49  
4                1     49,99    19,49

perlu dilakukan replace "," pada column price_usd dan cogs_usd dengan "."

In [ ]:
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32313 entries, 0 to 32312
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            32313 non-null  int64 
 1   created_at          32313 non-null  object
 2   website_session_id  32313 non-null  int64 
 3   user_id             32313 non-null  int64 
 4   primary_product_id  32313 non-null  int64 
 5   items_purchased     32313 non-null  int64 
 6   price_usd           32313 non-null  object
 7   cogs_usd            32313 non-null  object
dtypes: int64(5), object(3)
memory usage: 2.0+ MB


tidak ditemukan data di column yg bernilai Null

##CEK TIPE DATA

In [ ]:
df_orders.dtypes

order_id               int64
created_at            object
website_session_id     int64
user_id                int64
primary_product_id     int64
items_purchased        int64
price_usd             object
cogs_usd              object
dtype: object

tipe data yg perlu diubah yaitu 1. column created_at perlu diubah dr object ke datetime 2. column price_usd dan cogs_usd perlu diubah dari object ke float dan me replace "," menjadi "."

##CEK TYPO

In [ ]:
for column in df_orders:
    print(df_orders[column].value_counts())

order_id
32313    1
1        1
2        1
3        1
4        1
        ..
14       1
13       1
12       1
11       1
10       1
Name: count, Length: 32313, dtype: int64
created_at
12/15/14 15:28    4
9/17/14 15:03     3
7/16/14 11:15     3
11/28/14 14:18    3
11/28/14 9:48     3
                 ..
1/23/14 14:02     1
1/23/14 13:52     1
1/23/14 13:39     1
1/23/14 13:21     1
1/23/14 17:10     1
Name: count, Length: 31572, dtype: int64
website_session_id
472818    1
20        1
104       1
147       1
160       1
         ..
336       1
332       1
327       1
321       1
317       1
Name: count, Length: 32313, dtype: int64
user_id
68624     3
122640    3
276897    3
318660    3
275098    3
         ..
161883    1
161875    1
161857    1
161853    1
161940    1
Name: count, Length: 31696, dtype: int64
primary_product_id
1    23861
2     4803
3     3068
4      581
Name: count, dtype: int64
items_purchased
1    24601
2     7712
Name: count, dtype: int64
price_usd
49,99     18104
59,99

In [ ]:
df_orders['order_id'].value_counts()

order_id
32313    1
1        1
2        1
3        1
4        1
        ..
14       1
13       1
12       1
11       1
10       1
Name: count, Length: 32313, dtype: int64

setelah pengecekan dengan .value_counts() tdk ditemukan adanya TYPO di semua column pada table orders

##CEK DUPLICATE

In [ ]:
df_orders['order_id'].duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
32308    False
32309    False
32310    False
32311    False
32312    False
Name: order_id, Length: 32313, dtype: bool

In [ ]:
df_orders['created_at'].duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
32308    False
32309    False
32310    False
32311    False
32312    False
Name: created_at, Length: 32313, dtype: bool

In [ ]:
df_orders['website_session_id'].duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
32308    False
32309    False
32310    False
32311    False
32312    False
Name: website_session_id, Length: 32313, dtype: bool

In [ ]:
df_orders['user_id'].duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
32308    False
32309    False
32310    False
32311    False
32312    False
Name: user_id, Length: 32313, dtype: bool

In [ ]:
df_orders['primary_product_id'].duplicated()

0        False
1         True
2         True
3         True
4         True
         ...  
32308     True
32309     True
32310     True
32311     True
32312     True
Name: primary_product_id, Length: 32313, dtype: bool

In [ ]:
df_orders['items_purchased'].duplicated()

0        False
1         True
2         True
3         True
4         True
         ...  
32308     True
32309     True
32310     True
32311     True
32312     True
Name: items_purchased, Length: 32313, dtype: bool

In [ ]:
df_orders['price_usd'].duplicated()

0        False
1         True
2         True
3         True
4         True
         ...  
32308     True
32309     True
32310     True
32311     True
32312     True
Name: price_usd, Length: 32313, dtype: bool

In [ ]:
df_orders['cogs_usd'].duplicated()

0        False
1         True
2         True
3         True
4         True
         ...  
32308     True
32309     True
32310     True
32311     True
32312     True
Name: cogs_usd, Length: 32313, dtype: bool

setelah dilakukan pengecekan dengan .duplicated, tdk ditemukan adanya data yg duplicate di semua column pada tabel orders

##CEK MISSING VALUE

In [ ]:
df_orders.isnull().sum()

order_id              0
created_at            0
website_session_id    0
user_id               0
primary_product_id    0
items_purchased       0
price_usd             0
cogs_usd              0
dtype: int64

setelah dilakukan pengecekan missing value dengan .isnull().sum(), tidak ditemukan adanya data yg missing value di semua column pada tabel orders

##DATA CLEANING AND PREPARATION

###merubah tipe data di column created_at dari object menjadi datetime

In [ ]:
df_orders['created_at'] = pd.to_datetime(df_orders['created_at'], format='%m/%d/%y %H:%M')

In [ ]:
df_orders.dtypes

order_id                       int64
created_at            datetime64[ns]
website_session_id             int64
user_id                        int64
primary_product_id             int64
items_purchased                int64
price_usd                     object
cogs_usd                      object
dtype: object

In [ ]:
df_orders.head()

order_id          created_at  website_session_id  user_id  \
0         1 2012-03-19 10:42:00                  20       20   
1         2 2012-03-19 19:27:00                 104      104   
2         3 2012-03-20 06:44:00                 147      147   
3         4 2012-03-20 09:41:00                 160      160   
4         5 2012-03-20 11:28:00                 177      177   

   primary_product_id  items_purchased price_usd cogs_usd  
0                   1                1     49,99    19,49  
1                   1                1     49,99    19,49  
2                   1                1     49,99    19,49  
3                   1                1     49,99    19,49  
4                   1                1     49,99    19,49

###merubah tipe data column price_usd serta cogs_usd dari object ke float dan me replace "," menjadi "."

In [ ]:
df_orders['price_usd'] = df_orders['price_usd'].str.replace(',', '.').astype(float)
df_orders['cogs_usd'] = df_orders['cogs_usd'].str.replace(',', '.').astype(float)
print(df_orders[['price_usd', 'cogs_usd']].dtypes)

price_usd    float64
cogs_usd     float64
dtype: object


In [ ]:
df_orders.dtypes

order_id                       int64
created_at            datetime64[ns]
website_session_id             int64
user_id                        int64
primary_product_id             int64
items_purchased                int64
price_usd                    float64
cogs_usd                     float64
dtype: object

In [ ]:
df_orders.head()

order_id          created_at  website_session_id  user_id  \
0         1 2012-03-19 10:42:00                  20       20   
1         2 2012-03-19 19:27:00                 104      104   
2         3 2012-03-20 06:44:00                 147      147   
3         4 2012-03-20 09:41:00                 160      160   
4         5 2012-03-20 11:28:00                 177      177   

   primary_product_id  items_purchased  price_usd  cogs_usd  
0                   1                1      49.99     19.49  
1                   1                1      49.99     19.49  
2                   1                1      49.99     19.49  
3                   1                1      49.99     19.49  
4                   1                1      49.99     19.49

## #TABLE ORDER_ITEMS

In [ ]:
import pandas as pd

In [ ]:
potter_url = 'https://docs.google.com/spreadsheets/d/15E2af3_ABAN8vmQ2jext8fbr0H9MIHjl1suJhNE3rHI/edit?gid=1814474632#gid=1814474632'
spreadsheet_id = potter_url.split('/')[-2]
gid = potter_url.split('gid=')[-1].split('#')[0]
potter_csv_export_url = f'https://docs.google.com/spreadsheets/d/{spreadsheet_id}/export?format=csv&gid={gid}'
df_order_items = pd.read_csv(potter_csv_export_url)
df_order_items

order_item_id     created_at  order_id  product_id  is_primary_item  \
0                  1  3/19/12 10:42         1           1                1   
1                  2  3/19/12 19:27         2           1                1   
2                  3   3/20/12 6:44         3           1                1   
3                  4   3/20/12 9:41         4           1                1   
4                  5  3/20/12 11:28         5           1                1   
...              ...            ...       ...         ...              ...   
40020          40021   3/19/15 4:10     32310           4                1   
40021          40022   3/19/15 5:27     32311           2                1   
40022          40023   3/19/15 5:27     32311           4                0   
40023          40024   3/19/15 5:35     32312           4                1   
40024          40025   3/19/15 5:38     32313           1                1   

      price_usd cogs_usd  
0         49,99    19,49  
1         49,99    19,49  
2         49,99    19,49  
3         49,99    19,49  
4         49,99    19,49  
...         ...      ...  
40020     29,99     9,49  
40021     59,99    22,49  
40022     29,99     9,49  
40023     29,99     9,49  
40024     49,99    19,49  

[40025 rows x 7 columns]

In [ ]:
backup1=df_order_items.copy()

##CEK TIPE DATA

In [ ]:
df_order_items.dtypes

order_item_id       int64
created_at         object
order_id            int64
product_id          int64
is_primary_item     int64
price_usd          object
cogs_usd           object
dtype: object

tipe data yg perlu diubah yaitu 1. column created_at perlu diubah dr object ke datetime 2. column is_primary_item diubah dr interger menjadi boolean 3. column price_usd dan cogs_usd perlu diubah dari object ke float dan me replace "," menjadi "."

##CEK TYPO

In [ ]:
for column in df_order_items:
    print(df_order_items[column].value_counts())

order_item_id
40025    1
1        1
2        1
3        1
4        1
        ..
14       1
13       1
12       1
11       1
10       1
Name: count, Length: 40025, dtype: int64
created_at
3/18/14 11:50     6
12/15/14 15:28    6
12/1/14 10:36     6
12/1/14 13:12     5
7/16/14 11:15     5
                 ..
1/1/14 10:40      1
1/1/14 10:38      1
1/1/14 10:10      1
1/1/14 9:57       1
1/1/14 19:05      1
Name: count, Length: 31572, dtype: int64
order_id
25800    2
13082    2
20213    2
25797    2
25796    2
        ..
10112    1
10111    1
10110    1
10109    1
10124    1
Name: count, Length: 32313, dtype: int64
product_id
1    24226
2     5796
4     5018
3     4985
Name: count, dtype: int64
is_primary_item
1    32313
0     7712
Name: count, dtype: int64
price_usd
49,99    24226
59,99     5796
29,99     5018
45,99     4985
Name: count, dtype: int64
cogs_usd
19,49    24226
22,49     5796
9,49      5018
14,49     4985
Name: count, dtype: int64


In [ ]:
df_order_items['order_item_id'].value_counts()

order_item_id
40025    1
1        1
2        1
3        1
4        1
        ..
14       1
13       1
12       1
11       1
10       1
Name: count, Length: 40025, dtype: int64

In [ ]:
df_order_items['created_at'].value_counts()

created_at
3/18/14 11:50     6
12/15/14 15:28    6
12/1/14 10:36     6
12/1/14 13:12     5
7/16/14 11:15     5
                 ..
1/1/14 10:40      1
1/1/14 10:38      1
1/1/14 10:10      1
1/1/14 9:57       1
1/1/14 19:05      1
Name: count, Length: 31572, dtype: int64

In [ ]:
df_order_items['order_id'].value_counts()

order_id
25800    2
13082    2
20213    2
25797    2
25796    2
        ..
10112    1
10111    1
10110    1
10109    1
10124    1
Name: count, Length: 32313, dtype: int64

In [ ]:
df_order_items['product_id'].value_counts()

product_id
1    24226
2     5796
4     5018
3     4985
Name: count, dtype: int64

In [ ]:
df_order_items['is_primary_item'].value_counts()

is_primary_item
1    32313
0     7712
Name: count, dtype: int64

In [ ]:
df_order_items['price_usd'].value_counts()

price_usd
49,99    24226
59,99     5796
29,99     5018
45,99     4985
Name: count, dtype: int64

In [ ]:
df_order_items['cogs_usd'].value_counts()

cogs_usd
19,49    24226
22,49     5796
9,49      5018
14,49     4985
Name: count, dtype: int64

setelah pengecekan dengan .value_counts() tdk ditemukan adanya TYPO di semua column pada table order_items

##CEK DUPLICATE

In [ ]:
df_order_items['order_item_id'].duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
40020    False
40021    False
40022    False
40023    False
40024    False
Name: order_item_id, Length: 40025, dtype: bool

In [ ]:
df_order_items['created_at'].duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
40020    False
40021    False
40022     True
40023    False
40024    False
Name: created_at, Length: 40025, dtype: bool

In [ ]:
df_order_items['order_id'].duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
40020    False
40021    False
40022     True
40023    False
40024    False
Name: order_id, Length: 40025, dtype: bool

In [ ]:
df_order_items['product_id'].duplicated()

0        False
1         True
2         True
3         True
4         True
         ...  
40020     True
40021     True
40022     True
40023     True
40024     True
Name: product_id, Length: 40025, dtype: bool

In [ ]:
df_order_items['is_primary_item'].duplicated()

0        False
1         True
2         True
3         True
4         True
         ...  
40020     True
40021     True
40022     True
40023     True
40024     True
Name: is_primary_item, Length: 40025, dtype: bool

In [ ]:
df_order_items['price_usd'].duplicated()

0        False
1         True
2         True
3         True
4         True
         ...  
40020     True
40021     True
40022     True
40023     True
40024     True
Name: price_usd, Length: 40025, dtype: bool

In [ ]:
df_order_items['cogs_usd'].duplicated()

0        False
1         True
2         True
3         True
4         True
         ...  
40020     True
40021     True
40022     True
40023     True
40024     True
Name: cogs_usd, Length: 40025, dtype: bool

setelah dilakukan pengecekan dengan .duplicated, tdk ditemukan adanya data yg duplicate di semua column pada tabel order_items

##CEK MISSING VALUE

In [ ]:
df_order_items.isnull().sum()

order_item_id      0
created_at         0
order_id           0
product_id         0
is_primary_item    0
price_usd          0
cogs_usd           0
dtype: int64

setelah dilakukan pengecekan missing value dengan .isnull().sum(), tidak ditemukan adanya data yg missing value di semua column pada tabel order_items

##DATA CLEANING AND PREPARATION

###merubah tipe data di column created_at dari object menjadi datetime

In [ ]:
df_order_items['created_at'] = pd.to_datetime(df_order_items['created_at'], format='%m/%d/%y %H:%M')

In [ ]:
df_order_items.dtypes

order_item_id               int64
created_at         datetime64[ns]
order_id                    int64
product_id                  int64
is_primary_item             int64
price_usd                  object
cogs_usd                   object
dtype: object

In [ ]:
df_order_items.head()

order_item_id          created_at  order_id  product_id  is_primary_item  \
0              1 2012-03-19 10:42:00         1           1                1   
1              2 2012-03-19 19:27:00         2           1                1   
2              3 2012-03-20 06:44:00         3           1                1   
3              4 2012-03-20 09:41:00         4           1                1   
4              5 2012-03-20 11:28:00         5           1                1   

  price_usd cogs_usd  
0     49,99    19,49  
1     49,99    19,49  
2     49,99    19,49  
3     49,99    19,49  
4     49,99    19,49

###merubah tipe data di column is_primary_item dr interger menjadi boolean

In [ ]:
df_order_items['is_primary_item'] = df_order_items['is_primary_item'].astype(bool)

In [ ]:
df_order_items.dtypes

order_item_id               int64
created_at         datetime64[ns]
order_id                    int64
product_id                  int64
is_primary_item              bool
price_usd                  object
cogs_usd                   object
dtype: object

###merubah tipe data column price_usd serta cogs_usd dari object ke float dan me replace "," menjadi "."

In [ ]:
df_order_items['price_usd'] = df_order_items['price_usd'].str.replace(',', '.').astype(float)
df_order_items['cogs_usd'] = df_order_items['cogs_usd'].str.replace(',', '.').astype(float)
print(df_order_items[['price_usd', 'cogs_usd']].dtypes)

price_usd    float64
cogs_usd     float64
dtype: object


In [ ]:
df_order_items.head()

order_item_id          created_at  order_id  product_id  is_primary_item  \
0              1 2012-03-19 10:42:00         1           1             True   
1              2 2012-03-19 19:27:00         2           1             True   
2              3 2012-03-20 06:44:00         3           1             True   
3              4 2012-03-20 09:41:00         4           1             True   
4              5 2012-03-20 11:28:00         5           1             True   

   price_usd  cogs_usd  
0      49.99     19.49  
1      49.99     19.49  
2      49.99     19.49  
3      49.99     19.49  
4      49.99     19.49

## #TABLE ORDER_ITEM_REFUNDS

In [ ]:
import pandas as pd

In [ ]:
potter_url = 'https://docs.google.com/spreadsheets/d/1A2jHgdDj4FAGhu9JLS5tXEv80xNv4zKfYGpVMMd2qVQ/edit?gid=968260780#gid=968260780'
spreadsheet_id = potter_url.split('/')[-2]
gid = potter_url.split('gid=')[-1].split('#')[0]
potter_csv_export_url = f'https://docs.google.com/spreadsheets/d/{spreadsheet_id}/export?format=csv&gid={gid}'
df_order_item_refunds = pd.read_csv(potter_csv_export_url)
df_order_item_refunds

order_item_refund_id     created_at  order_item_id  order_id  \
0                        1   4/6/12 11:32             57        57   
1                        2   4/13/12 1:09             74        74   
2                        3   4/15/12 7:03             71        71   
3                        4  4/17/12 20:00            118       118   
4                        5  4/22/12 20:53            116       116   
...                    ...            ...            ...       ...   
1726                  1727   3/30/15 9:37          39950     32255   
1727                  1728  3/30/15 21:33          39671     32049   
1728                  1729  3/31/15 19:59          39729     32090   
1729                  1730    4/1/15 3:54          39717     32079   
1730                  1731   4/1/15 18:11          39947     32252   

     refund_amount_usd  
0                49,99  
1                49,99  
2                49,99  
3                49,99  
4                49,99  
...                ...  
1726             59,99  
1727             49,99  
1728             49,99  
1729             59,99  
1730             45,99  

[1731 rows x 5 columns]

In [ ]:
backup1=df_order_item_refunds.copy()

##CEK TIPE DATA

In [ ]:
df_order_item_refunds.dtypes

order_item_refund_id     int64
created_at              object
order_item_id            int64
order_id                 int64
refund_amount_usd       object
dtype: object

tipe data yg perlu diubah yaitu 1. column created_at perlu diubah dr object ke datetime 2. column refund_amount_usd perlu diubah dari object ke float dan me replace "," menjadi "."

##CEK TYPO

In [ ]:
for column in df_order_item_refunds:
    print(df_order_item_refunds[column].value_counts())

order_item_refund_id
1731    1
1       1
2       1
1715    1
1714    1
       ..
8       1
7       1
6       1
5       1
4       1
Name: count, Length: 1731, dtype: int64
created_at
9/5/14 2:37      2
9/7/14 7:47      2
9/19/14 21:54    1
9/19/14 18:29    1
9/19/14 15:08    1
                ..
2/14/14 9:38     1
2/14/14 9:10     1
2/14/14 8:29     1
2/12/14 3:28     1
2/23/14 12:02    1
Name: count, Length: 1729, dtype: int64
order_item_id
39947    1
57       1
74       1
39813    1
39172    1
        ..
191      1
186      1
147      1
116      1
118      1
Name: count, Length: 1731, dtype: int64
order_id
23321    2
24472    2
9324     2
27061    2
28802    2
        ..
11124    1
11157    1
10904    1
11084    1
11265    1
Name: count, Length: 1723, dtype: int64
refund_amount_usd
49,99    1237
45,99     301
59,99     129
29,99      64
Name: count, dtype: int64


setelah pengecekan dengan .value_counts() tdk ditemukan adanya TYPO di semua column pada table order_item_refunds

##CEK DUPLICATE

In [ ]:
df_order_item_refunds['order_item_refund_id'].duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1726    False
1727    False
1728    False
1729    False
1730    False
Name: order_item_refund_id, Length: 1731, dtype: bool

In [ ]:
df_order_item_refunds['created_at'].duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1726    False
1727    False
1728    False
1729    False
1730    False
Name: created_at, Length: 1731, dtype: bool

In [ ]:
df_order_item_refunds['order_item_id'].duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1726    False
1727    False
1728    False
1729    False
1730    False
Name: order_item_id, Length: 1731, dtype: bool

In [ ]:
df_order_item_refunds['order_id'].duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1726    False
1727    False
1728    False
1729    False
1730    False
Name: order_id, Length: 1731, dtype: bool

In [ ]:
df_order_item_refunds['refund_amount_usd'].duplicated()

0       False
1        True
2        True
3        True
4        True
        ...  
1726     True
1727     True
1728     True
1729     True
1730     True
Name: refund_amount_usd, Length: 1731, dtype: bool

setelah dilakukan pengecekan dengan .duplicated, tdk ditemukan adanya data yg duplicate di semua column pada tabel order_item_refunds

##CEK MISSING VALUE

In [ ]:
df_order_item_refunds.isnull().sum()

order_item_refund_id    0
created_at              0
order_item_id           0
order_id                0
refund_amount_usd       0
dtype: int64

setelah dilakukan pengecekan missing value dengan .isnull().sum(), tidak ditemukan adanya data yg missing value di semua column pada tabel order_item_refunds

##DATA CLEANING AND PREPARATION

###merubah tipe data di column created_at dari object menjadi datetime

In [ ]:
df_order_item_refunds['created_at'] = pd.to_datetime(df_order_item_refunds['created_at'], format='%m/%d/%y %H:%M')

In [ ]:
df_order_item_refunds.dtypes

order_item_refund_id             int64
created_at              datetime64[ns]
order_item_id                    int64
order_id                         int64
refund_amount_usd               object
dtype: object

In [ ]:
df_order_item_refunds.head()

order_item_refund_id          created_at  order_item_id  order_id  \
0                     1 2012-04-06 11:32:00             57        57   
1                     2 2012-04-13 01:09:00             74        74   
2                     3 2012-04-15 07:03:00             71        71   
3                     4 2012-04-17 20:00:00            118       118   
4                     5 2012-04-22 20:53:00            116       116   

  refund_amount_usd  
0             49,99  
1             49,99  
2             49,99  
3             49,99  
4             49,99

###merubah tipe data di column refund_amount_usd perlu dari object ke float dan me replace "," menjadi "."

In [ ]:
df_order_item_refunds['refund_amount_usd'] = df_order_item_refunds['refund_amount_usd'].str.replace(',', '.').astype(float)

In [ ]:
df_order_item_refunds.dtypes

order_item_refund_id             int64
created_at              datetime64[ns]
order_item_id                    int64
order_id                         int64
refund_amount_usd              float64
dtype: object

In [ ]:
df_order_item_refunds.head()

order_item_refund_id          created_at  order_item_id  order_id  \
0                     1 2012-04-06 11:32:00             57        57   
1                     2 2012-04-13 01:09:00             74        74   
2                     3 2012-04-15 07:03:00             71        71   
3                     4 2012-04-17 20:00:00            118       118   
4                     5 2012-04-22 20:53:00            116       116   

   refund_amount_usd  
0              49.99  
1              49.99  
2              49.99  
3              49.99  
4              49.99

##MERGED DATA (orders and order_items)

In [ ]:
df_merged = pd.merge(df_orders, df_order_items, on='order_id', how='inner')
display(df_merged.head())

order_id        created_at_x  website_session_id  user_id  \
0         1 2012-03-19 10:42:00                  20       20   
1         2 2012-03-19 19:27:00                 104      104   
2         3 2012-03-20 06:44:00                 147      147   
3         4 2012-03-20 09:41:00                 160      160   
4         5 2012-03-20 11:28:00                 177      177   

   primary_product_id  items_purchased  price_usd_x  cogs_usd_x  \
0                   1                1        49.99       19.49   
1                   1                1        49.99       19.49   
2                   1                1        49.99       19.49   
3                   1                1        49.99       19.49   
4                   1                1        49.99       19.49   

   order_item_id        created_at_y  product_id  is_primary_item  \
0              1 2012-03-19 10:42:00           1             True   
1              2 2012-03-19 19:27:00           1             True   
2              3 2012-03-20 06:44:00           1             True   
3              4 2012-03-20 09:41:00           1             True   
4              5 2012-03-20 11:28:00           1             True   

   price_usd_y  cogs_usd_y  
0        49.99       19.49  
1        49.99       19.49  
2        49.99       19.49  
3        49.99       19.49  
4        49.99       19.49